### ISSUE TITLE
run_experiment

### ISSUE THREAD

**Issue by Annashuvarina24** [Annashuvarina24/Evaluation-metrics#14]
def run_experiment(df, target_column, cv_strategies=["expanding", "rolling", "gap", "seasonal", "multi_horizon"]):
    """
    Runs an experiment using an LSTM-based model with different Cross-Validation strategies.
    Logs results for comparison.

    Parameters:
        df (DataFrame): The dataset with a datetime index.
        target_column (str): The target variable.
        cv_strategies (list): List of cross-validation strategies to test.

    Returns:
        DataFrame containing all CV results.
    """
    logger.info(f"🚀 Running experiment for {target_column} across multiple CV strategies...")

    try:
        # ✅ Ensure Datetime Index
        if not isinstance(df.index, pd.DatetimeIndex):
            df.index = pd.to_datetime(df.index)

        # ✅ Preprocess Data with STL and Feature Engineering
        df = stl_preprocess(df, target_column)
        df = add_features(df, target_column)

        if df.empty:
            logger.error(f"❌ DataFrame is empty after feature engineering for {target_column}. Skipping experiment.")
            return None

        results = []  # Store results for each CV strategy

        # =======================
        # 🛠️ TEST DIFFERENT CV STRATEGIES
        # =======================
        for cv_strategy in cv_strategies:
            logger.info(f"📌 Testing {cv_strategy} Cross-Validation...")

            # ✅ Extract features & target
            X_train_full = df.drop(columns=[target_column])
            y_train_full = df[target_column]

            # ✅ Scale Features & Target
            scaler_X, scaler_y = StandardScaler(), StandardScaler()
            X_train_full_scaled = scaler_X.fit_transform(X_train_full)
            y_train_full_scaled = scaler_y.fit_transform(y_train_full.values.reshape(-1, 1)).flatten()

            # ✅ Ensure Proper Shapes
            if len(X_train_full_scaled) != len(y_train_full_scaled):
                logger.error("❌ Feature and target size mismatch! Skipping this experiment.")
                continue

            # ✅ Generate Model Inside Loop (For Isolation)
            model = Sequential([
                Input(shape=(X_train_full.shape[1],)),
                Dense(256, activation='relu'),
                Dropout(0.2),
                Dense(128, activation='relu'),
                Dropout(0.2),
                Dense(64, activation='relu'),
                Dense(1)
            ])
            model.compile(optimizer=Adam(learning_rate=0.001, clipnorm=1.0), loss='mse')

            mean_rmse = cross_validate_with_timeseries_split(
                model=model,
                X_train=pd.DataFrame(X_train_full_scaled, index=df.index, columns=X_train_full.columns),
                y_train=pd.Series(y_train_full_scaled, index=df.index),
                strategy=cv_strategy
            )

            # ✅ Store results for comparison
            results.append({
                "Target": target_column,
                "CV Strategy": cv_strategy,
                "RMSE": mean_rmse
            })

        # ✅ Convert to DataFrame for better visualization
        results_df = pd.DataFrame(results)
        logger.info(f"🏆 CV Comparison for {target_column}:\n{results_df}")

        # =======================
        # 🏁 FINAL TRAINING & TESTING
        # =======================
        logger.info(f"📌 Training final model on full training data for {target_column}...")

        train_data = df.loc[CONFIG['train_start']:CONFIG['train_end']]
        test_data = df.loc[CONFIG['test_start']:CONFIG['test_end']]

        if train_data.empty or test_data.empty:
            logger.error(f"❌ Training or test data is empty. Cannot proceed with final model.")
            return results_df

        X_train_final, y_train_final = train_data.drop(columns=[target_column]), train_data[target_column]
        X_test, y_test = test_data.drop(columns=[target_column]), test_data[target_column]

        # ✅ Scale Final Train & Test Data
        X_train_final_scaled = scaler_X.fit_transform(X_train_final)
        X_test_scaled = scaler_X.transform(X_test)
        y_train_final_scaled = scaler_y.fit_transform(y_train_final.values.reshape(-1, 1)).flatten()

        # ✅ Define Final Model (Isolated for Separate Training)
        model = Sequential([
            Input(shape=(X_train_final_scaled.shape[1],)),
            Dense(256, activation='relu'),
            Dropout(0.2),
            Dense(128, activation='relu'),
            Dropout(0.2),
            Dense(64, activation='relu'),
            Dense(1)
        ])
        model.compile(optimizer=Adam(learning_rate=0.001, clipnorm=1.0), loss='mse')

        # ✅ Implement Early Stopping
        early_stopping = EarlyStopping(monitor='loss', patience=5, restore_best_weights=True)

        # ✅ Train Final Model on Full Data
        model.fit(X_train_final_scaled, y_train_final_scaled, epochs=50, batch_size=32, verbose=1, callbacks=[early_stopping])

        # ✅ Make Predictions on Test Set
        y_test_pred_scaled = model.predict(X_test_scaled)
        y_test_pred = scaler_y.inverse_transform(y_test_pred_scaled).flatten()

        # ✅ Evaluate Final Test Performance
        final_mae, final_mae_percent, final_rmse, final_rmse_percent, final_mape, final_smape = evaluate_metrics(y_test, y_test_pred)

        logger.info(f"🏁 Final Test Results for {target_column}: MAE: {final_mae:.4f}, RMSE: {final_rmse:.4f}, "
                    f"MAPE: {final_mape:.2f}%, SMAPE: {final_smape:.2f}%")

        # ✅ Store final model results
        final_results = {
            "Target": target_column,
            "CV Strategy": "Final Test",
            "RMSE": final_rmse,
            "MAE": final_mae,
            "MAPE": final_mape,
            "SMAPE": final_smape
        }
        final_results_df = pd.DataFrame([final_results])

        return pd.concat([results_df, final_results_df], ignore_index=True)

    except Exception as e:
        logger.error(f"❌ Experiment for {target_column} failed: {e}")
        return None

# Load Reactive and Active Power Data
reactive_power_path = r'C:\Users\912125\PycharmProjects\pythonProject\XGBoost related files\processed_tr-r_q102.csv'
active_power_path = r'C:\Users\912125\PycharmProjects\pythonProject\XGBoost related files\processed_tr-r_p102.csv'
weather_path = r'C:\Users\912125\PycharmProjects\pythonProject\XGBoost related files\STN_370_modified.csv'

try :
    # Load reactive and active power datasets
    reactive_df = pd.read_csv(reactive_power_path)
    active_df = pd.read_csv(active_power_path)

    # Load weather dataset and filter relevant columns
    weather_df = pd.read_csv(weather_path)
    weather_df = weather_df[ [ 'Date', 'Time', 't', 'ff' ] ]  # Keep only relevant columns

    # Rename weather columns for clarity
    weather_df.rename(columns={
        't' : 'temperature',
        'ff' : 'wind_speed',
    }, inplace=True)

    # Parse datetime and set as index for all datasets
    reactive_df[ 'datetime' ] = pd.to_datetime(reactive_df[ 'Date' ] + ' ' + reactive_df[ 'Time' ], errors='coerce')
    active_df[ 'datetime' ] = pd.to_datetime(active_df[ 'Date' ] + ' ' + active_df[ 'Time' ], errors='coerce')
    weather_df[ 'datetime' ] = pd.to_datetime(weather_df[ 'Date' ] + ' ' + weather_df[ 'Time' ], errors='coerce')

    reactive_df.set_index('datetime', inplace=True)
    active_df.set_index('datetime', inplace=True)
    weather_df.set_index('datetime', inplace=True)

    # Drop unnecessary columns
    reactive_df.drop(columns=['Date', 'Time'], inplace=True)
    active_df.drop(columns=['Date', 'Time'], inplace=True)
    weather_df.drop(columns=['Date', 'Time'], inplace=True)
    files = [reactive_power_path, active_power_path, weather_path]
    for file in files:
        if not os.path.exists(file):
            logger.error(f"❌ ERROR: File not found: {file}")

    # Debug: Print the first few rows of each dataset
    print(" Reactive Power Data Sample:\n", reactive_df.head())
    print(" Active Power Data Sample:\n", active_df.head())
    print(" Weather Data Sample:\n", weather_df.head())

    # Check for missing values
    print(" Missing Values in Reactive Power Data:\n", reactive_df.isnull().sum())
    print(" Missing Values in Active Power Data:\n", active_df.isnull().sum())
    print(" Missing Values in Weather Data:\n", weather_df.isnull().sum())

    # Check for datetime parsing errors
    print(" Reactive Power Datetime Sample:\n", reactive_df.index[ :5 ])
    print(" Active Power Datetime Sample:\n", active_df.index[ :5 ])
    print(" Weather Datetime Sample:\n", weather_df.index[ :5 ])

    # ✅ Merge datasets on datetime
    df = reactive_df.join(active_df, rsuffix="_active").join(weather_df, rsuffix="_weather")

    # ✅ Ensure the index has a **fixed hourly frequency**
    df = df.asfreq("h")  # 🔹 Use lowercase "h" to prevent warning

    # ✅ Check if any timestamps are missing
    missing_timestamps = df.index.to_series().diff().value_counts()
    print("Time Differences:", missing_timestamps)

    # ✅ Fix Deprecated Warning (Use .ffill() and .bfill() instead)
    df.ffill(inplace=True)  # ✅ Forward-fill missing values
    df.bfill(inplace=True)  # ✅ Backward-fill missing values

    # If still missing values exist, fill with the column mean
    df.fillna(df.mean(), inplace=True)

    # ✅ Check & Remove Duplicates
    if df.index.duplicated().sum() > 0 :
        df = df[ ~df.index.duplicated(keep="first") ]
        logger.warning("⚠️ Duplicate timestamps detected and removed!")

    logger.info("✅ Datasets successfully loaded, merged, and resampled to hourly frequency.")

except Exception as e :
    logger.error(f"❌ Error loading data: {e}")
    exit(1)

all_results = []

# ✅ Ensure target columns exist in df and are numeric
if not isinstance(df.index, pd.DatetimeIndex):
    df.index = pd.to_datetime(df.index)  # Ensure datetime index

valid_target_columns = [
    col for col in reactive_df.columns
    if col in df.columns and np.issubdtype(df[col].dtype, np.number)
]

if not valid_target_columns:
    logger.warning("⚠️ No valid numeric target columns found in dataset. Exiting.")
else:
    for target_column in valid_target_columns:
        try:
            logger.info(f"🟢 Running experiment for {target_column}...")
            results_df = run_experiment(df, target_column)

            if results_df is not None and not results_df.empty:
                all_results.append(results_df)
            else:
                logger.warning(f"⚠️ Experiment for {target_column} did not return valid results.")
        except Exception as e:
            logger.error(f"❌ Experiment failed for {target_column}: {e}")

# ✅ Combine results into a final DataFrame
if all_results:
    final_results_df = pd.concat(all_results, ignore_index=True)

    # ✅ Display final results in a UI-friendly format
    import ace_tools as tools
    tools.display_dataframe_to_user(name="Cross-Validation Results", dataframe=final_results_df)

    # ✅ Save results to CSV
    results_path = os.path.join(CONFIG["results_dir"], "cv_results.csv")
    final_results_df.to_csv(results_path, index=False)
    logger.info(f"✅ Cross-validation results saved to {results_path}")

else:
    logger.warning("⚠️ No valid results generated. Check experiment logs.")

logger.info(f"Script completed in {time.time() - script_start_time:.2f} seconds.")
